blockchain/state_migration.go

```diff
@@ -371,3 +372,132 @@ func (bc *BlockChain) StopStateMigration() error {
 func (bc *BlockChain) StatusStateMigration() (bool, uint64, int, int, float64) {
 	return bc.db.InMigration(), bc.db.MigrationBlockNumber(), bc.committedCnt, bc.pendingCnt, bc.progress
 }
+
+func (bc *BlockChain) concurrentIterateTrie(root common.Hash, db state.Database, resultCh chan common.Hash, finishCh chan error) (resultErr error) {
+	defer func() {
+		finishCh <- resultErr
+	}()
+
+	stateDB, err := state.New(root, db)
+	if err != nil {
+		return err
+	}
+
+	it := state.NewNodeIterator(stateDB)
+	for it.Next() {
+		resultCh <- it.Hash
+
+		select {
+		case <-bc.quitWarmUp:
+			return errors.New("quitWarmUp")
+		case <-bc.quit:
+			return errors.New("quit")
+		default:
+		}
+	}
+
+	return nil
+}
+
+func (bc *BlockChain) warmUpLoop(cache *fastcache.Cache, mainTrieCacheLimit uint64, children []common.Hash, resultHashCh chan common.Hash, resultErrCh chan error) {
+	logged := time.Now()
+	var context []interface{}
+	var stats fastcache.Stats
+	var percent uint64
+	var cnt int
+
+	updateContext := func() {
+		stats = fastcache.Stats{}
+		cache.UpdateStats(&stats)
+		percent = stats.BytesSize * 100 / mainTrieCacheLimit
+		context = []interface{}{
+			"warmUpCnt", cnt,
+			"cacheLimit", units.Base2Bytes(mainTrieCacheLimit).String(),
+			"cachedSize", units.Base2Bytes(stats.BytesSize).String(),
+			"percent", percent,
+		}
+	}
+
+	var resultErr error
+	for finishCnt := 0; finishCnt < len(children); {
+		select {
+		case <-resultHashCh:
+			cnt++
+			if time.Since(logged) > log.StatsReportLimit {
+				logged = time.Now()
+
+				updateContext()
+				if percent > 90 { //more than 90%
+					close(bc.quitWarmUp)
+					logger.Info("Warm up is completed", context...)
+					return
+				}
+
+				logger.Info("Warm up progress", context...)
+			}
+		case err := <-resultErrCh:
```

Comment:
How about renaming the channel to `finishCh`? It is confusing when it is `nil`.

```suggestion
		case err := <-finishCh:
```

Reply:
I added a comment below. :)